src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBulkCopy.cs

```diff
@@ -1395,22 +1544,192 @@ public void WriteToServer(DbDataReader reader)
+
+                WriteRowSourceToServerAsync(table.Columns.Count, CancellationToken.None); //It returns null since _isAsyncBulkCopy = false;
             }
             finally
             {
                 SqlStatistics.StopTimer(statistics);
             }
         }
 
+        public void WriteToServer(DataRow[] rows)
+        {
+            SqlStatistics statistics = Statistics;
+
+            if (rows == null)
+            {
+                throw new ArgumentNullException(nameof(rows));
+            }
+
+            if (_isBulkCopyingInProgress)
+            {
+                throw SQL.BulkLoadPendingOperation();
+            }
+
+            if (rows.Length == 0)
+            {
+                return; // nothing to do. user passed us an empty array
+            }
+
+            try
+            {
+                statistics = SqlStatistics.StartTimer(Statistics);
+
+                DataTable table = rows[0].Table;
+                Debug.Assert(null != table, "How can we have rows without a table?");
+                _rowStateToSkip = DataRowState.Deleted;      // Don't allow deleted rows
+                _rowSource = rows;
+                _dataTableSource = table;
+                _SqlDataReaderRowSource = null;
+                _rowSourceType = ValueSourceType.RowArray;
+                _rowEnumerator = rows.GetEnumerator();
+                _isAsyncBulkCopy = false;
+
+                WriteRowSourceToServerAsync(table.Columns.Count, CancellationToken.None); //It returns null since _isAsyncBulkCopy = false;
+            }
+            finally
+            {
+                SqlStatistics.StopTimer(statistics);
+            }
+        }
+
+        public Task WriteToServerAsync(DataRow[] rows)
+        {
+            return WriteToServerAsync(rows, CancellationToken.None);
+        }
+
+        public Task WriteToServerAsync(DataRow[] rows, CancellationToken cancellationToken)
+        {
+            Task resultTask = null;
+
+            if (rows == null)
+            {
+                throw new ArgumentNullException(nameof(rows));
+            }
+
+            if (_isBulkCopyingInProgress)
+            {
+                throw SQL.BulkLoadPendingOperation();
+            }
+
+            SqlStatistics statistics = Statistics;
+            try
+            {
+                statistics = SqlStatistics.StartTimer(Statistics);
+
+                if (rows.Length == 0)
+                {
+                    TaskCompletionSource<object> source = new TaskCompletionSource<object>();
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        source.SetCanceled();
+                    }
+                    else
+                    {
+                        source.SetResult(null);
+                    }
+                    resultTask = source.Task;
+                    return resultTask; // nothing to do. user passed us an empty array. Return a completed Task.
+                }
```

Comment:
The `TaskCompletionSource<T>` allocation can be avoided and the above can be simplified as:

```c#
if (rows.Length == 0)
{
    return cancellationToken.IsCancellationRequested ?
        Task.FromCanceled(cancellationToken) :
        Task.CompletedTask;
}
```